// dayjs-nuxt@1.2.7/dist/module.mjs
import { defineNuxtModule, createResolver, addPlugin, addImports, addTemplate } from '@nuxt/kit';
import { resolve } from 'pathe';

const module = defineNuxtModule({
  meta: {
    name: "dayjs",
    configKey: "dayjs",
    compatibility: {
      nuxt: "^3"
    }
  },
  // Default configuration options of the Nuxt module
  defaults: {
    locales: [],
    plugins: ["updateLocale", "utc"],
    defaultLocale: void 0,
    defaultTimezone: void 0
  },
  async setup(options, nuxt) {
    const resolver = createResolver(import.meta.url);
    options.plugins = [...new Set(options.plugins)];
    if (options.defaultTimezone && !options.plugins.includes("timezone"))
      throw new Error("You must include the timezone plugin in order to set a default timezone");
    const runtimeDir = await resolver.resolve("./runtime");
    nuxt.options.build.transpile.push(runtimeDir);
    addPlugin(resolve(runtimeDir, "plugin"));
    addImports({
      name: "useDayjs",
      as: "useDayjs",
      from: resolve(runtimeDir, "composables")
    });
    addTemplate({
      filename: "dayjs.imports.mjs",
      getContents: () => generateImports(options)
    });
    nuxt.hook("prepare:types", ({ references }) => {
      if (options.plugins) {
        const plugins = options.plugins.map((p) => ({ types: `dayjs/plugin/${p}` }));
        references.push(...plugins);
      }
    });
  }
});
const generateImports = ({ locales, plugins, defaultLocale, defaultTimezone }) => `
// Generated by dayjs-nuxt-module
import dayjs from 'dayjs'
${locales?.map((locale) => `import 'dayjs/locale/${locale}'`).join("\n")}
${plugins?.map((plugin) => `import ${plugin} from 'dayjs/plugin/${plugin}'`).join("\n")}

${plugins?.map((plugin) => `dayjs.extend(${plugin})`).join("\n")}
${defaultLocale ? "dayjs.extend(updateLocale)" : ""}
${locales?.map((locale) => `dayjs.locale('${locale}')`).join("\n")}
${defaultTimezone ? `dayjs.tz.setDefault('${defaultTimezone}')` : ""}

${defaultLocale ? `
dayjs.updateLocale(${JSON.stringify(defaultLocale).replace(/^\[|\]$/g, "")})` : ""}

export default dayjs
`;

export { module as default };
